### PR TITLE
Add mod validation on user and room changes

### LINKED ISSUE #45: Mod compatibility needs to be checked server-side
Basically we need an implementation [similar to osu-web](https://github.com/ppy/osu-web/blob/05ca44bcb6f0b99b1a12f2923131599498aa65e1/app/Libraries/Multiplayer/Mod.php) to ensure correct mods are chosen by the client.

This is already enforced client side but we shouldn't be relying on that forever.

### PR DESCRIPTION
**Pull request by peppy**
- Closes #45.
- Closes ppy/osu#13567 (as far as I've checked once the server is authoritative this should stop all crashes)

### CODE REVIEW

**Comment by peppy** [comment]
We probably also want checks that all `UserMods` are in `AllowedMods`. I'll either do that in this PR or a follow-up depending on review status.

**Review by bdach — CHANGES_REQUESTED** [review]

**Comment by bdach** [comment]
I've already touched on this in one of the comments above, but depending on how paranoid we want to be, the next step would be to verify the validity of `AllowedMods` and `RequiredMods` themselves server-side, too.

File: osu.Server.Spectator.Tests/MultiplayerFlowTests.cs [hunk 1]
```diff
@@ -786,6 +789,243 @@ public async Task OnlyClientsInSameRoomReceiveAvailabilityChange()
...
+            await hub.JoinRoom(room_id);
+
+            await hub.ChangeSettings(new MultiplayerRoomSettings
+            {
+                RulesetID = 2,
+                BeatmapChecksum = "checksum",
+                AllowedMods = new[]
+                {
+                    new APIMod(new CatchModHidden()),
+                    new APIMod(new OsuModApproachDifferent())
```
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Server.Spectator.Tests/MultiplayerFlowTests.cs:912]: definitely for a separate PR, as this one is focused only on `UserMods`.

**Comment by peppy** [comment]
> I've already touched on this in one of the comments above, but depending on how paranoid we want to be, the next step would be to verify the validity of `AllowedMods` and `RequiredMods` themselves server-side, too.

I'll do that in a separate PR today, I think. This one should still be fine to be merged on its own as an improvement over what we have, regardless.

**Review by smoogipoo — APPROVED** [review]
lgtm

**Pull request merged**